packages/unlock-js/src/PublicLock/v4/purchaseKeys.js

```diff
@@ -0,0 +1,45 @@
+import purchaseKey from './purchaseKey'
+
+/**
+ * Purchase key function. This implementation requires the following
+ * @param {object} params:
+ * - {PropTypes.arrayOf(PropTypes.address)} lockAddress
+ * - {PropTypes.arrayOf(PropTypes.address)} owners
+ * - {PropTypes.arrayOf(string)} keyPrices
+ * - {PropTypes.address} erc20Address
+ * - {number} decimals
+ * - {PropTypes.arrayOf(PropTypes.address)} referrers (address which will receive UDT - if applicable)
+ * - {PropTypes.arrayOf(PropTypes.array[bytes])} _data (array of array of bytes, not used in transaction but can be used by hooks)
+ * @param {function} callback invoked with the transaction hash
+ */
+export default async function (
+  {
+    lockAddress,
+    erc20Address,
+    decimals,
+    owners = [],
+    keyManagers = [],
+    keyPrices = [],
+    referrers = [],
+    data = [],
+  },
+  callback
+) {
+  return await Promise.all(
+    owners.map(async (owner, i) =>
```

Comment:
I had to move that logic to a wrapper because for ERC20 the allowance call was failing so we need to approve the entire amount before calling `purchaseKey`. Works properly now :)